<!DOCTYPE html>
<html data-color-mode="light" data-dark-theme="dark" data-light-theme="light" lang="zh-CN">
<head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <link href='https://mirrors.sustech.edu.cn/cdnjs/ajax/libs/Primer/21.0.7/primer.css' rel='stylesheet' />
    
    <link rel="icon" href="https://avatars.githubusercontent.com/u/103297347?v=4"><script>
        let theme = localStorage.getItem("meek_theme") || "light";
        document.documentElement.setAttribute("data-color-mode", theme);
    </script>
<meta name="description" content="## COSG是什么

**COSG**(COSine similarity-based marker Gene identification)是由来自哈佛医学院和Broad研究所博后Ming Dai等人开发，旨在从余弦相似度的角度鉴定cluster的marker gene，于2021年12月被Briefings in Bioinformatics接收。">
<meta property="og:title" content="单细胞-COSG鉴定marker基因">
<meta property="og:description" content="## COSG是什么

**COSG**(COSine similarity-based marker Gene identification)是由来自哈佛医学院和Broad研究所博后Ming Dai等人开发，旨在从余弦相似度的角度鉴定cluster的marker gene，于2021年12月被Briefings in Bioinformatics接收。">
<meta property="og:type" content="article">
<meta property="og:url" content="https://bionoob7.github.io/post/dan-xi-bao--COSG-jian-ding-marker-ji-yin.html">
<meta property="og:image" content="https://avatars.githubusercontent.com/u/103297347?v=4">
<title>单细胞-COSG鉴定marker基因</title>
<link href="//unpkg.com/@wooorm/starry-night@2.1.1/style/both.css" rel="stylesheet" />


</head>
<style>
body{box-sizing: border-box;min-width: 200px;max-width: 900px;margin: 20px auto;padding: 45px;font-size: 16px;font-family: sans-serif;line-height: 1.25;}
#header{display:flex;padding-bottom:8px;border-bottom: 1px solid var(--borderColor-muted, var(--color-border-muted));margin-bottom: 16px;}
#footer {margin-top:64px; text-align: center;font-size: small;}

</style>

<style>
.postTitle{margin: auto 0;font-size:40px;font-weight:bold;}
.title-right{display:flex;margin:auto 0 0 auto;}
.title-right .circle{padding: 14px 16px;margin-right:8px;}
#postBody{border-bottom: 1px solid var(--color-border-default);padding-bottom:36px;}
#postBody hr{height:2px;}
#cmButton{height:48px;margin-top:48px;}
#comments{margin-top:64px;}
.g-emoji{font-size:24px;}
@media (max-width: 600px) {
    body {padding: 8px;}
    .postTitle{font-size:24px;}
}
.copy-feedback {
    display: none;
    position: absolute;
    top: 10px;
    right: 50px;
    color: var(--color-fg-on-emphasis);
    background-color: var(--color-fg-muted);
    border-radius: 3px;
    padding: 5px 8px;
    font-size: 12px;
}
</style>
<style>.markdown-alert{padding:0.5rem 1rem;margin-bottom:1rem;border-left:.25em solid var(--borderColor-default,var(--color-border-default));}.markdown-alert .markdown-alert-title {display:flex;font-weight:var(--base-text-weight-medium,500);align-items:center;line-height:1;}.markdown-alert>:first-child {margin-top:0;}.markdown-alert>:last-child {margin-bottom:0;}</style><style>.markdown-alert.markdown-alert-note {border-left-color:var(--borderColor-accent-emphasis, var(--color-accent-emphasis));background-color:var(--color-accent-subtle);}.markdown-alert.markdown-alert-note .markdown-alert-title {color: var(--fgColor-accent,var(--color-accent-fg));}</style>



<body>
    <div id="header">
<h1 class="postTitle">单细胞-COSG鉴定marker基因</h1>
<div class="title-right">
    <a href="https://bionoob7.github.io" id="buttonHome" class="btn btn-invisible circle" title="首页">
        <svg class="octicon" width="16" height="16">
            <path id="pathHome" fill-rule="evenodd"></path>
        </svg>
    </a>
    
    <a href="https://github.com/bionoob7/bionoob7.github.io/issues/19" target="_blank" class="btn btn-invisible circle" title="Issue">
        <svg class="octicon" width="16" height="16">
            <path id="pathIssue" fill-rule="evenodd"></path>
        </svg>
    </a>
    

    <a class="btn btn-invisible circle" onclick="modeSwitch();" title="切换主题">
        <svg class="octicon" width="16" height="16" >
            <path id="themeSwitch" fill-rule="evenodd"></path>
        </svg>
    </a>

</div>
</div>
    <div id="content">
<div class="markdown-body" id="postBody"><h2>COSG是什么</h2>
<p><strong>COSG</strong>(COSine similarity-based marker Gene identification)是由来自哈佛医学院和Broad研究所博后Ming Dai等人开发，旨在从余弦相似度的角度鉴定cluster的marker gene，于2021年12月被Briefings in Bioinformatics接收。目前已将此方法分别包装为R包与Python包，分别对应Seurat与Scanny分析流程。</p>
<ul>
<li>
<p>Paper：<a href="https://doi.org/10.1093/bib/bbab579" rel="nofollow">https://doi.org/10.1093/bib/bbab579</a></p>
</li>
<li>
<p>Github：<a href="https://github.com/genecell/COSG%EF%BC%8C">https://github.com/genecell/COSG，</a> <a href="https://github.com/genecell/COSGR">https://github.com/genecell/COSGR</a></p>
</li>
</ul>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701094817968.png"><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701094817968.png" alt="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701094817968.png" style="max-width: 100%; height: auto;"></a></p>
<h2>为什么选择COSG</h2>
<p>传统的marker基因鉴定方法基于欧式距离，即目标细胞类群与其它细胞之间的基因表达差异来进行marker基因的鉴定。但这样的鉴定方法存在潜在的问题：当marker基因在目标细胞类群以外的其它一小部分细胞中高表达时（如某个小的细胞亚型），传统的鉴定方法仍然会将其鉴定为marker基因，但因为小范围的高表达对于整体的细胞群来说不具备决定性意义，所以仍然会认为这个marker基因在其余的细胞群中是低表达的。而COSG（COSine similarity-based marker Gene identification），即基于余弦值的marker基因鉴定方法可以解决上述问题。原因在于余弦值不依赖于向量的模，放到单细胞分析的背景下就是不依赖于基因的表达量，而依赖于基因的表达模式。举例为证：基因A在细胞1和2中的表达量分别是2和4，基因B在细胞1和2中的表达量分别是1和2，如果根据欧式距离的算法，显然基因A和基因B会被定义成为不同的表达模式，但是如果我们基于余弦值的话，两个基因在该细胞集合中的表达向量之间的夹角就为：</p>
<p><a target="_blank" rel="noopener noreferrer" href="https://github.com/user-attachments/assets/4f2b6c32-00ee-4268-807c-8628fb4c0042"><img src="https://github.com/user-attachments/assets/4f2b6c32-00ee-4268-807c-8628fb4c0042" alt="e42712dc0303045064c81ce5e270caa6" style="max-width: 100%; height: auto;"></a></p>
<p>在这个背景下，两个基因表达向量之间的夹角为0，也就是这两个基因在这个细胞集合中的表达模式实际上是相同的。</p>
<h2>怎么做COSG分析</h2>
<h3>COSG分析思路</h3>
<p>（1）准备好数据， 即标准化后的单细胞表达矩阵，并完成细胞聚类分析(scATAC， ST同理)；</p>
<p>（2）对cluster K模拟一个ideal marker gene(仅在k表达，其余cluster不表达)；</p>
<p>（3）计算所有真实基因分别与该基因的余弦相似度，并进一步计算出COSG score；</p>
<p>（4）COSG score取值范围为0~1。<strong>值越大，表示为cluster specific marker gene</strong>；</p>
<p>（5）同理依次计算出其余cluster的marker gene。</p>
<h3>与其他方法的性能比较</h3>
<p>文章主要与常用的Wilcoxon-test等分析方式在多种数据集上进行比较，概括如下：</p>
<p>（1）在模拟的单细胞数据集中，COSG方法可以最大程度发现每个clutser的marker gene；</p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100442863.png"><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100442863.png" alt="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100442863.png" style="max-width: 100%; height: auto;"></a></p>
<p>（2）在大规模单细胞数据中， COSG在保持精度的同时分析速度很快；</p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100648514.png"><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100648514.png" alt="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100648514.png" style="max-width: 100%; height: auto;"></a></p>
<p>（3）COSG在scATAC以及空间转录组数据上也同样具有较优的表现</p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100846995.png"><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100846995.png" alt="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701100846995.png" style="max-width: 100%; height: auto;"></a></p>
<h3>代码实操</h3>
<div class="highlight highlight-source-r"><pre class="notranslate"><span class="pl-e">remotes</span><span class="pl-k">::</span>install_github(<span class="pl-v">repo</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>genecell/COSGR<span class="pl-pds">'</span></span>)

library(<span class="pl-smi">COSG</span>)
library(<span class="pl-smi">Seurat</span>)

<span class="pl-c"><span class="pl-c">#</span># 输入数据</span>
data(<span class="pl-s"><span class="pl-pds">'</span>pbmc_small<span class="pl-pds">'</span></span>,<span class="pl-v">package</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>COSG<span class="pl-pds">'</span></span>)

dim(<span class="pl-smi">pbmc_small</span><span class="pl-k">@</span><span class="pl-smi">assays</span><span class="pl-k">$</span><span class="pl-smi">RNA</span><span class="pl-k">@</span><span class="pl-smi">data</span>)  <span class="pl-c"><span class="pl-c">#</span>标准化矩阵</span>
<span class="pl-c"><span class="pl-c">#</span> [1] 230  80</span>

table(Idents(<span class="pl-smi">pbmc_small</span>)) <span class="pl-c"><span class="pl-c">#</span>聚类分群结果</span>
<span class="pl-c"><span class="pl-c">#</span> 0  1  2 </span>
<span class="pl-c"><span class="pl-c">#</span> 36 25 19 </span></pre></div>
<ul>
<li>R包的分析函数即为<code class="notranslate">cosg()</code>，默认参数如下</li>
</ul>
<div class="highlight highlight-source-r"><pre class="notranslate"><span class="pl-smi">marker_cosg</span> <span class="pl-k">&lt;-</span> cosg(
  <span class="pl-smi">pbmc_small</span>,
  <span class="pl-v">groups</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>all<span class="pl-pds">'</span></span>, <span class="pl-c"><span class="pl-c">#</span>考虑全部分组</span>
  <span class="pl-v">assay</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>RNA<span class="pl-pds">'</span></span>,
  <span class="pl-v">slot</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>data<span class="pl-pds">'</span></span>,
  <span class="pl-v">mu</span><span class="pl-k">=</span><span class="pl-c1">1</span>,         <span class="pl-c"><span class="pl-c">#</span>惩罚项参数，值越大</span>
  <span class="pl-v">remove_lowly_expressed</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>,   <span class="pl-c"><span class="pl-c">#</span>是否过滤低表达基因</span>
  <span class="pl-v">expressed_pct</span><span class="pl-k">=</span><span class="pl-c1">0.1</span>,             <span class="pl-c"><span class="pl-c">#</span>设置低表达的阈值</span>
  <span class="pl-v">n_genes_user</span><span class="pl-k">=</span><span class="pl-c1">100</span>      <span class="pl-c"><span class="pl-c">#</span>每个cluster定义Top-N个marker gene</span>
)

head(<span class="pl-smi">marker_cosg</span><span class="pl-k">$</span><span class="pl-smi">names</span>,<span class="pl-c1">3</span>)
<span class="pl-c"><span class="pl-c">#</span>      0      1     2</span>
<span class="pl-c"><span class="pl-c">#</span> 1  CD7 S100A8 MS4A1</span>
<span class="pl-c"><span class="pl-c">#</span> 2 CCL5   TYMP CD79A</span>
<span class="pl-c"><span class="pl-c">#</span> 3 GNLY S100A9 TCL1A</span>
head(<span class="pl-smi">marker_cosg</span><span class="pl-k">$</span><span class="pl-smi">scores</span>,<span class="pl-c1">3</span>)
<span class="pl-c"><span class="pl-c">#</span>           0         1         2</span>
<span class="pl-c"><span class="pl-c">#</span> 1 0.6391917 0.8954042 0.6922908</span>
<span class="pl-c"><span class="pl-c">#</span> 2 0.6391267 0.8312083 0.5832425</span>
<span class="pl-c"><span class="pl-c">#</span> 3 0.6328148 0.8120045 0.5757478</span></pre></div>
<ul>
<li>简单可视化一下</li>
</ul>
<div class="highlight highlight-source-r"><pre class="notranslate"><span class="pl-smi">top_list</span><span class="pl-k">&lt;-</span>c()
<span class="pl-k">for</span> (<span class="pl-smi">group</span> <span class="pl-k">in</span> colnames(<span class="pl-smi">marker_cosg</span><span class="pl-k">$</span><span class="pl-smi">names</span>)){
  <span class="pl-smi">top_i</span><span class="pl-k">&lt;-</span><span class="pl-smi">marker_cosg</span><span class="pl-k">$</span><span class="pl-smi">names</span>[<span class="pl-smi">group</span>][<span class="pl-c1">1</span><span class="pl-k">:</span><span class="pl-c1">5</span>,<span class="pl-c1">1</span>]
  <span class="pl-smi">top_list</span><span class="pl-k">&lt;-</span>c(<span class="pl-smi">top_list</span>,<span class="pl-smi">top_i</span>)
}

DotPlot(<span class="pl-smi">pbmc_small</span>, 
        <span class="pl-v">assay</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>RNA<span class="pl-pds">'</span></span>,
        <span class="pl-c"><span class="pl-c">#</span> scale=TRUE,</span>
        <span class="pl-v">features</span> <span class="pl-k">=</span>  unique(<span class="pl-smi">top_list</span>)) <span class="pl-k">+</span> RotatedAxis()

<span class="pl-smi">pbmc_small</span> <span class="pl-k">&lt;-</span> ScaleData(<span class="pl-smi">pbmc_small</span>, <span class="pl-v">features</span> <span class="pl-k">=</span> <span class="pl-smi">top_list</span>)
DoHeatmap(<span class="pl-smi">pbmc_small</span>, 
          <span class="pl-v">assay</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>RNA<span class="pl-pds">'</span></span>,
          <span class="pl-v">features</span> <span class="pl-k">=</span>  <span class="pl-smi">top_list</span>) </pre></div>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/b01b959552e760093fbbf8d08fc0490067f4ac0e4c266ef365d71b6f790b37a2/68747470733a2f2f74632d63646e2e666c6f7775732e636e2f6f73732f66353066353462382d303233392d343466342d396432662d3930656238383838343765392f696d6167652e706e673f74696d653d3137323832303136303026746f6b656e3d3633353833616266366130363335363938333839326238356163326363353635343463653865356336623264666464356230393562653661343561396632643226726f6c653d66726565"><img src="https://camo.githubusercontent.com/b01b959552e760093fbbf8d08fc0490067f4ac0e4c266ef365d71b6f790b37a2/68747470733a2f2f74632d63646e2e666c6f7775732e636e2f6f73732f66353066353462382d303233392d343466342d396432662d3930656238383838343765392f696d6167652e706e673f74696d653d3137323832303136303026746f6b656e3d3633353833616266366130363335363938333839326238356163326363353635343463653865356336623264666464356230393562653661343561396632643226726f6c653d66726565" alt="image.png" data-canonical-src="https://tc-cdn.flowus.cn/oss/f50f54b8-0239-44f4-9d2f-90eb888847e9/image.png?time=1728201600&amp;token=63583abf6a06356983892b85ac2cc56544ce8e5c6b2dfdd5b095be6a45a9f2d2&amp;role=free" style="max-width: 100%; height: auto;"></a></p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701102521289.png"><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701102521289.png" alt="https://raw.githubusercontent.com/lishensuo/images/main/image-20230701102521289.png" style="max-width: 100%; height: auto;"></a></p>
<h2>tips</h2>
<div class="markdown-alert markdown-alert-note"><p class="markdown-alert-title"><svg class="octicon octicon-info mr-2" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Note</p><p>正常NFS推荐使用该方法，作者不推荐在SCTransform之后使用该方法</p>
</div>
<blockquote>
<p>The above error might be caused by using the SCT assay, which is not sparse after the SCTransform normalization. We recommend use the <code class="notranslate">RNA</code> assay with the slot <code class="notranslate">data</code> as the input.</p>
</blockquote></div>
<div style="font-size:small;margin-top:8px;float:right;"></div>

<button class="btn btn-block" type="button" onclick="openComments()" id="cmButton">评论</button>
<div class="comments" id="comments"></div>

</div>
    <div id="footer"><div id="footer1">Copyright © <span id="copyrightYear"></span> <a href="https://bionoob7.github.io">Ma's bioinfo notebook</a></div>
<div id="footer2">
    <span id="runday"></span><span>Powered by <a href="https://meekdai.com/Gmeek.html" target="_blank">Gmeek</a></span>
</div>

<script>
var now=new Date();
document.getElementById("copyrightYear").innerHTML=now.getFullYear();

if(""!=""){
    var startSite=new Date("");
    var diff=now.getTime()-startSite.getTime();
    var diffDay=Math.floor(diff/(1000*60*60*24));
    document.getElementById("runday").innerHTML="网站运行"+diffDay+"天"+" • ";
}
</script></div>
</body>
<script>
var IconList={'sun': 'M8 10.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5zM8 12a4 4 0 100-8 4 4 0 000 8zM8 0a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0V.75A.75.75 0 018 0zm0 13a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0v-1.5A.75.75 0 018 13zM2.343 2.343a.75.75 0 011.061 0l1.06 1.061a.75.75 0 01-1.06 1.06l-1.06-1.06a.75.75 0 010-1.06zm9.193 9.193a.75.75 0 011.06 0l1.061 1.06a.75.75 0 01-1.06 1.061l-1.061-1.06a.75.75 0 010-1.061zM16 8a.75.75 0 01-.75.75h-1.5a.75.75 0 010-1.5h1.5A.75.75 0 0116 8zM3 8a.75.75 0 01-.75.75H.75a.75.75 0 010-1.5h1.5A.75.75 0 013 8zm10.657-5.657a.75.75 0 010 1.061l-1.061 1.06a.75.75 0 11-1.06-1.06l1.06-1.06a.75.75 0 011.06 0zm-9.193 9.193a.75.75 0 010 1.06l-1.06 1.061a.75.75 0 11-1.061-1.06l1.06-1.061a.75.75 0 011.061 0z', 'moon': 'M9.598 1.591a.75.75 0 01.785-.175 7 7 0 11-8.967 8.967.75.75 0 01.961-.96 5.5 5.5 0 007.046-7.046.75.75 0 01.175-.786zm1.616 1.945a7 7 0 01-7.678 7.678 5.5 5.5 0 107.678-7.678z', 'sync': 'M1.705 8.005a.75.75 0 0 1 .834.656 5.5 5.5 0 0 0 9.592 2.97l-1.204-1.204a.25.25 0 0 1 .177-.427h3.646a.25.25 0 0 1 .25.25v3.646a.25.25 0 0 1-.427.177l-1.38-1.38A7.002 7.002 0 0 1 1.05 8.84a.75.75 0 0 1 .656-.834ZM8 2.5a5.487 5.487 0 0 0-4.131 1.869l1.204 1.204A.25.25 0 0 1 4.896 6H1.25A.25.25 0 0 1 1 5.75V2.104a.25.25 0 0 1 .427-.177l1.38 1.38A7.002 7.002 0 0 1 14.95 7.16a.75.75 0 0 1-1.49.178A5.5 5.5 0 0 0 8 2.5Z', 'home': 'M6.906.664a1.749 1.749 0 0 1 2.187 0l5.25 4.2c.415.332.657.835.657 1.367v7.019A1.75 1.75 0 0 1 13.25 15h-3.5a.75.75 0 0 1-.75-.75V9H7v5.25a.75.75 0 0 1-.75.75h-3.5A1.75 1.75 0 0 1 1 13.25V6.23c0-.531.242-1.034.657-1.366l5.25-4.2Zm1.25 1.171a.25.25 0 0 0-.312 0l-5.25 4.2a.25.25 0 0 0-.094.196v7.019c0 .138.112.25.25.25H5.5V8.25a.75.75 0 0 1 .75-.75h3.5a.75.75 0 0 1 .75.75v5.25h2.75a.25.25 0 0 0 .25-.25V6.23a.25.25 0 0 0-.094-.195Z', 'github': 'M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z', 'copy': 'M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z', 'check': 'M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z'};
var utterancesLoad=0;

let themeSettings={
    "dark": ["dark","moon","#00f0ff","dark-blue"],
    "light": ["light","sun","#ff5000","github-light"],
    "auto": ["auto","sync","","preferred-color-scheme"]
};
function changeTheme(mode, icon, color, utheme){
    document.documentElement.setAttribute("data-color-mode",mode);
    document.getElementById("themeSwitch").setAttribute("d",value=IconList[icon]);
    document.getElementById("themeSwitch").parentNode.style.color=color;
    if(utterancesLoad==1){utterancesTheme(utheme);}
}
function modeSwitch(){
    let currentMode=document.documentElement.getAttribute('data-color-mode');
    let newMode = currentMode === "light" ? "dark" : currentMode === "dark" ? "auto" : "light";
    localStorage.setItem("meek_theme", newMode);
    if(themeSettings[newMode]){
        changeTheme(...themeSettings[newMode]);
    }
}
function utterancesTheme(theme){
    const message={type:'set-theme',theme: theme};
    const iframe=document.getElementsByClassName('utterances-frame')[0];
    iframe.contentWindow.postMessage(message,'https://utteranc.es');
}
if(themeSettings[theme]){changeTheme(...themeSettings[theme]);}
console.log("\n %c Gmeek last https://github.com/Meekdai/Gmeek \n","padding:5px 0;background:#02d81d;color:#fff");
</script>

<script>
document.getElementById("pathHome").setAttribute("d",IconList["home"]);
document.getElementById("pathIssue").setAttribute("d",IconList["github"]);



function openComments(){
    cm=document.getElementById("comments");
    cmButton=document.getElementById("cmButton");
    cmButton.innerHTML="loading";
    span=document.createElement("span");
    span.setAttribute("class","AnimatedEllipsis");
    cmButton.appendChild(span);

    script=document.createElement("script");
    script.setAttribute("src","https://utteranc.es/client.js");
    script.setAttribute("repo","bionoob7/bionoob7.github.io");
    script.setAttribute("issue-term","title");
    
    if(localStorage.getItem("meek_theme")=="dark"){script.setAttribute("theme","dark-blue");}
    else if(localStorage.getItem("meek_theme")=="light") {script.setAttribute("theme","github-light");}
    else{script.setAttribute("theme","preferred-color-scheme");}
    
    script.setAttribute("crossorigin","anonymous");
    script.setAttribute("async","");
    cm.appendChild(script);

    int=self.setInterval("iFrameLoading()",200);
}

function iFrameLoading(){
    var utterances=document.getElementsByClassName('utterances');
    if(utterances.length==1){
        if(utterances[0].style.height!=""){
            utterancesLoad=1;
            int=window.clearInterval(int);
            document.getElementById("cmButton").style.display="none";
            console.log("utterances Load OK");
        }
    }
}

document.addEventListener('DOMContentLoaded', () => {
    const createClipboardHTML = (codeContent, additionalClasses = '') => `
        <pre class="notranslate"><code class="notranslate">${codeContent}</code></pre>
        <div class="clipboard-container position-absolute right-0 top-0 ${additionalClasses}">
            <clipboard-copy class="ClipboardButton btn m-2 p-0" role="button" style="display: inherit;">
                <svg height="16" width="16" class="octicon octicon-copy m-2"><path d="${IconList["copy"]}"></path></svg>
                <svg height="16" width="16" class="octicon octicon-check color-fg-success m-2 d-none"><path d="${IconList["check"]}"></path></svg>
            </clipboard-copy>
            <div class="copy-feedback">Copied!</div>
        </div>
    `;

    const handleCodeElements = (selector = '') => {
        document.querySelectorAll(selector).forEach(codeElement => {
            const codeContent = codeElement.innerHTML;
            const newStructure = document.createElement('div');
            newStructure.className = 'snippet-clipboard-content position-relative overflow-auto';
            newStructure.innerHTML = createClipboardHTML(codeContent);

            const parentElement = codeElement.parentElement;
            if (selector.includes('highlight')) {
                parentElement.insertBefore(newStructure, codeElement.nextSibling);
                parentElement.removeChild(codeElement);
            } else {
                parentElement.parentElement.replaceChild(newStructure, parentElement);
            }
        });
    };

    handleCodeElements('pre.notranslate > code.notranslate');
    handleCodeElements('div.highlight > pre.notranslate');

    let currentFeedback = null;
    document.querySelectorAll('clipboard-copy').forEach(copyButton => {
        copyButton.addEventListener('click', () => {
            const codeContent = copyButton.closest('.snippet-clipboard-content').innerText;
            const tempTextArea = document.createElement('textarea');
            tempTextArea.value = codeContent;
            document.body.appendChild(tempTextArea);
            tempTextArea.select();
            document.execCommand('copy');
            document.body.removeChild(tempTextArea);

            const copyIcon = copyButton.querySelector('.octicon-copy');
            const checkIcon = copyButton.querySelector('.octicon-check');
            const copyFeedback = copyButton.nextElementSibling;

            if (currentFeedback && currentFeedback !== copyFeedback) {currentFeedback.style.display = 'none';}
            currentFeedback = copyFeedback;

            copyIcon.classList.add('d-none');
            checkIcon.classList.remove('d-none');
            copyFeedback.style.display = 'block';
            copyButton.style.borderColor = 'var(--color-success-fg)';

            setTimeout(() => {
                copyIcon.classList.remove('d-none');
                checkIcon.classList.add('d-none');
                copyFeedback.style.display = 'none';
                copyButton.style.borderColor = '';
            }, 2000);
        });
    });
});

</script>
<script src='https://blog.meekdai.com/Gmeek/plugins/GmeekTOC.js'></script><script src='https://blog.meekdai.com/Gmeek/plugins/lightbox.js'></script>

</html>
