<!DOCTYPE html>
<html data-color-mode="light" data-dark-theme="dark" data-light-theme="light" lang="zh-CN">
<head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <link href='https://mirrors.sustech.edu.cn/cdnjs/ajax/libs/Primer/21.0.7/primer.css' rel='stylesheet' />
    
    <link rel="icon" href="https://avatars.githubusercontent.com/u/103297347?v=4"><script>
        let theme = localStorage.getItem("meek_theme") || "light";
        document.documentElement.setAttribute("data-color-mode", theme);
    </script>
<meta name="description" content="# fasta

是一种基于文本用于表示核酸序列或多肽序列的格式。">
<meta property="og:title" content="生物信息学常见数据格式">
<meta property="og:description" content="# fasta

是一种基于文本用于表示核酸序列或多肽序列的格式。">
<meta property="og:type" content="article">
<meta property="og:url" content="https://bionoob7.github.io/post/sheng-wu-xin-xi-xue-chang-jian-shu-ju-ge-shi.html">
<meta property="og:image" content="https://avatars.githubusercontent.com/u/103297347?v=4">
<title>生物信息学常见数据格式</title>
<link href="//unpkg.com/@wooorm/starry-night@2.1.1/style/both.css" rel="stylesheet" />


</head>
<style>
body{box-sizing: border-box;min-width: 200px;max-width: 900px;margin: 20px auto;padding: 45px;font-size: 16px;font-family: sans-serif;line-height: 1.25;}
#header{display:flex;padding-bottom:8px;border-bottom: 1px solid var(--borderColor-muted, var(--color-border-muted));margin-bottom: 16px;}
#footer {margin-top:64px; text-align: center;font-size: small;}

</style>

<style>
.postTitle{margin: auto 0;font-size:40px;font-weight:bold;}
.title-right{display:flex;margin:auto 0 0 auto;}
.title-right .circle{padding: 14px 16px;margin-right:8px;}
#postBody{border-bottom: 1px solid var(--color-border-default);padding-bottom:36px;}
#postBody hr{height:2px;}
#cmButton{height:48px;margin-top:48px;}
#comments{margin-top:64px;}
.g-emoji{font-size:24px;}
@media (max-width: 600px) {
    body {padding: 8px;}
    .postTitle{font-size:24px;}
}
.copy-feedback {
    display: none;
    position: absolute;
    top: 10px;
    right: 50px;
    color: var(--color-fg-on-emphasis);
    background-color: var(--color-fg-muted);
    border-radius: 3px;
    padding: 5px 8px;
    font-size: 12px;
}
</style>




<body>
    <div id="header">
<h1 class="postTitle">生物信息学常见数据格式</h1>
<div class="title-right">
    <a href="https://bionoob7.github.io" id="buttonHome" class="btn btn-invisible circle" title="首页">
        <svg class="octicon" width="16" height="16">
            <path id="pathHome" fill-rule="evenodd"></path>
        </svg>
    </a>
    
    <a href="https://github.com/bionoob7/bionoob7.github.io/issues/29" target="_blank" class="btn btn-invisible circle" title="Issue">
        <svg class="octicon" width="16" height="16">
            <path id="pathIssue" fill-rule="evenodd"></path>
        </svg>
    </a>
    

    <a class="btn btn-invisible circle" onclick="modeSwitch();" title="切换主题">
        <svg class="octicon" width="16" height="16" >
            <path id="themeSwitch" fill-rule="evenodd"></path>
        </svg>
    </a>

</div>
</div>
    <div id="content">
<div class="markdown-body" id="postBody"><h1>fasta</h1>
<p>是一种基于文本用于表示核酸序列或多肽序列的格式。其中核酸或复基酸均以单个字母来表示，且允许在序列前添加序列名及注释</p>
<pre lang="Plain" data-meta="Text" class="notranslate"><code class="notranslate">&gt;MCHU - Calmodulin - Human, rabbit, bovine, rat, and chicken
MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID
FPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREA
DIDGDGQVNYEEFVQMMTAK*
</code></pre>
<p>特征：2行，id行和序列行。</p>
<ul>
<li>
<p>id行以“&gt;”开头，有时候会包含注释信息</p>
</li>
<li>
<p>序列行一个字母表示一个碱基（A、T、C、G、N）/氨基酸（20种常见氨基酸）</p>
</li>
</ul>
<h1>fastq</h1>
<p>是一种存储了生物序列以及相应的质量评价的文本格式。测序的原始序列</p>
<pre lang="Plain" data-meta="Text" class="notranslate"><code class="notranslate">@SEQ_ID
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65
</code></pre>
<p>特征：4行为一个单位</p>
<ul>
<li>
<p>第一行，id行，以@开头，记录必要信息</p>
</li>
<li>
<p>第二行，序列行</p>
</li>
<li>
<p>第三行，附加信息行(一般就一个加号)</p>
</li>
<li>
<p>第四行，碱基质量行</p>
</li>
</ul>
<h1>gff/gtf</h1>
<h2>gff(General Feature Format)</h2>
<p>记录序列中转录起始位点、基因、外显子、内含子等组成元件在染色体中的位置信息<br>
现在用得比较多的是第3版，即gff3</p>
<markdown-accessiblity-table><table role="table">
<thead>
<tr>
<th>Position index</th>
<th>Position name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>seqid</td>
<td>序列的编号。通常是染色体或者Contig ID，比如：Chr01 或者 scaffold_1</td>
</tr>
<tr>
<td>2</td>
<td>source</td>
<td>注释的来源。一般会是预测用软件工具或者数据库</td>
</tr>
<tr>
<td>3</td>
<td>type</td>
<td>The feature type name, like "gene" or "exon". In a well structured GFF file, all the children features always follow their parents in a single block (so all exons of a transcript are put after their parent "transcript" feature line and before any other parent transcript line). In GFF3, all features and their relationships should be compatible with the <a href="http://www.sequenceontology.org/gff3.shtml" rel="nofollow">standards released by the Sequence Ontology Project</a>.</td>
</tr>
<tr>
<td>4</td>
<td>start</td>
<td>Genomic start of the feature, with a <strong>1-base offset</strong>. This is in contrast with other 0-offset half-open sequence formats, like <a href="https://en.wikipedia.org/wiki/BED_(file_format)" rel="nofollow">BED</a>.</td>
</tr>
<tr>
<td>5</td>
<td>end</td>
<td>Genomic end of the feature, with a <strong>1-base offset</strong>. This is the same end coordinate as it is in 0-offset half-open sequence formats, like <a href="https://en.wikipedia.org/wiki/BED_(file_format)" rel="nofollow">BED</a>.[<em><a href="https://en.wikipedia.org/wiki/Wikipedia:Citation_needed" rel="nofollow">citation needed</a></em>]</td>
</tr>
<tr>
<td>6</td>
<td>score</td>
<td>Numeric value that generally indicates the confidence of the source in the annotated feature. A value of "." (a dot) is used to define a null value.</td>
</tr>
<tr>
<td>7</td>
<td>strand</td>
<td>Single character that indicates the <a href="https://en.wikipedia.org/wiki/Sense_(molecular_biology)#DNA_sense" rel="nofollow">strand</a> of the feature. This can be "+" (positive, or 5'-&gt;3'), "-", (negative, or 3'-&gt;5'), "." (undetermined), or "?" for features with relevant but unknown strands.</td>
</tr>
<tr>
<td>8</td>
<td>phase</td>
<td>phase of CDS features; it can be either one of 0, 1, 2 (for CDS features) or "." (for everything else). See the section below for a detailed explanation.</td>
</tr>
<tr>
<td>9</td>
<td>attributes</td>
<td>A list of tag-value pairs separated by a semicolon with additional information about the feature.</td>
</tr>
</tbody>
</table></markdown-accessiblity-table>
<pre lang="Plain" data-meta="Text" class="notranslate"><code class="notranslate">
##gff-version 3
# This is a test sample
scaffold625	maker	gene	337818	343277	.	+	.	ID=CLUHARG00000005458;Name=TUBB3_2
scaffold625	maker	mRNA	337818	343277	.	+	.	ID=CLUHART00000008717;Parent=CLUHARG00000005458
scaffold625	maker	tss	337916	337918	.	+	.	ID=CLUHART00000008717:tss;Parent=CLUHART00000008717
scaffold625	maker	start_codon	337916	337918	.	+	.	ID=CLUHART00000008717:start;Parent=CLUHART00000008717
scaffold625	maker	CDS	337915	337971	.	+	0	ID=CLUHART00000008717:cds;Parent=CLUHART00000008717
scaffold625	maker	CDS	340733	340841	.	+	0	ID=CLUHART00000008717:cds;Parent=CLUHART00000008717
scaffold625	maker	CDS	341518	341628	.	+	2	ID=CLUHART00000008717:cds;Parent=CLUHART00000008717
scaffold625	maker	CDS	341964	343033	.	+	2	ID=CLUHART00000008717:cds;Parent=CLUHART00000008717
scaffold625	maker	stop_codon	343031	343033	.	+	.	ID=CLUHART00000008717:stop;Parent=CLUHART00000008717
scaffold625	maker	exon	337818	337971	.	+	.	ID=CLUHART00000008717:exon1;Parent=CLUHART00000008717
scaffold625	maker	exon	340733	340841	.	+	.	ID=CLUHART00000008717:exon2;Parent=CLUHART00000008717
scaffold625	maker	exon	341518	341628	.	+	.	ID=CLUHART00000008717:exon3;Parent=CLUHART00000008717
scaffold625	maker	exon	341964	343277	.	+	.	ID=CLUHART00000008717:exon4;Parent=CLUHART00000008717
</code></pre>
<h2>gtf(General Transfer Format)</h2>
<p>现在用得比较多的是第2版，即gtf2</p>
<p>可以用cuff1inks里的gffreadi命令互相转换格式</p>
<p>格式：</p>
<ul>
<li>
<p><code class="notranslate">seqname</code>： 序列名称。通常是染色体或者Contig ID，比如：Chr01 或者 scaffold_1</p>
</li>
<li>
<p><code class="notranslate">source</code>：注释的来源。一般会是预测用软件工具或者数据库</p>
</li>
<li>
<p><code class="notranslate">feature</code>：注释信息类型（基因结构），这列信息是必须的，至少包括比如CDS、start_codon、stop_codon，还可以包含5UTR、3UTR、inter、inter_CNS、intron_CNS、exon、mRNA等等</p>
</li>
<li>
<p><code class="notranslate">start</code>：上述<code class="notranslate">feature</code>起始位点。从1 起始计数</p>
</li>
<li>
<p><code class="notranslate">end</code>：上述<code class="notranslate">feature</code>结束位点</p>
</li>
<li>
<p><code class="notranslate">score</code>：表示<code class="notranslate">feature</code>的存在和坐标的置信度，可以是一个浮点数或者整数，"." 表示为空，就是不需要</p>
</li>
<li>
<p><code class="notranslate">strand</code>：该<code class="notranslate">feature</code>位于参考序列的正链（+）或者负链（+）</p>
</li>
<li>
<p><code class="notranslate">frame</code>：0 表示阅读框的第一个完整密码子位于最5’端，1 表示在第一个完整密码子之前有一个额外的碱基，2 表示在第一整个密码子之前还有两个额外的碱基。<strong>注意</strong>，<code class="notranslate">frame</code>不是CDS长度除以3的余数（CDS mod 3）。如果链是“-”，则该区域的第一个碱基的值为<code class="notranslate">end</code>，因为对应的编码区域将在反向链上从<code class="notranslate">end</code>到<code class="notranslate">start</code>。</p>
</li>
<li>
<p><code class="notranslate">attributes</code> ：应具有的格式为<code class="notranslate">attribute_name “attribut_value”；</code>，每个属性必须以分号结尾并且与下一个属性之间以空格分隔，并且属性的值应该用双引号包围。并且<strong>必须包含</strong>以下两个属性：</p>
</li>
</ul>
<markdown-accessiblity-table><table role="table">
<thead>
<tr>
<th>attributes</th>
<th>含义</th>
</tr>
</thead>
<tbody>
<tr>
<td>gene_id "value";</td>
<td>表示转录本在基因组上的基因座的唯一ID。gene_id与value值用空格分开，如果值为空，则表示没有对应的基因。</td>
</tr>
<tr>
<td>transcript_id "value";</td>
<td>预测转录本的唯一ID。transcript_id与value值用空格分开，空表示没有转录本。</td>
</tr>
</tbody>
</table></markdown-accessiblity-table>
<pre lang="Plain" data-meta="Text" class="notranslate"><code class="notranslate">140    Twinscan    inter         5141     8522     .    -    .    gene_id ""; transcript_id "";
140    Twinscan    inter_CNS     8523     9711     .    -    .    gene_id ""; transcript_id "";
140    Twinscan    inter         9712     13182    .    -    .    gene_id ""; transcript_id "";
140    Twinscan    3UTR          65149    65487    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    3UTR          66823    66992    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    stop_codon    66993    66995    .    -    0    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    CDS           66996    66999    .    -    1    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    intron_CNS    70103    70151    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    CDS           70207    70294    .    -    2    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    CDS           71696    71807    .    -    0    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    start_codon   71805    71806    .    -    0    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    start_codon   73222    73222    .    -    2    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    CDS           73222    73222    .    -    0    gene_id "140.000"; transcript_id "140.000.1";
140    Twinscan    5UTR          73223    73504    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
</code></pre>
<p>转换</p>
<div class="highlight highlight-source-shell"><pre class="notranslate"><span class="pl-c"><span class="pl-c">#</span>from GFF to GTF</span>
gffread example.gff3 -T -o new.gtf
<span class="pl-c"><span class="pl-c">#</span>from GTF to GFF</span>
gffread example.gtf -o- <span class="pl-k">&gt;</span> new.gff3</pre></div>
<p>但是，使用gffread进行格式转换时容易导致信息的丢失，因此更推荐使用另一种软件AGAT。AGAT不仅可以进行GFF/GTF文件的格式转换，还能够检查修复、填充GFF和GTF的缺失信息。</p>
<p>AGAT</p>
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/74d514c5557ba55b2e6a42904e77c7a85b042aeef0e9f5c6aeb7fd22839a1769/68747470733a2f2f74632d63646e2e666c6f7775732e636e2f6f73732f37343132623666312d346465312d343637312d383865632d3934373333633139623331362f31353831333832616330343132323063316565376230396237646565646138612e706e673f74696d653d3137323934383638303026746f6b656e3d6464323433663731363234316531303666303833353662633637663364346332646333376333613930633036626533336535386335383839376139616661396426726f6c653d66726565"><img src="https://camo.githubusercontent.com/74d514c5557ba55b2e6a42904e77c7a85b042aeef0e9f5c6aeb7fd22839a1769/68747470733a2f2f74632d63646e2e666c6f7775732e636e2f6f73732f37343132623666312d346465312d343637312d383865632d3934373333633139623331362f31353831333832616330343132323063316565376230396237646565646138612e706e673f74696d653d3137323934383638303026746f6b656e3d6464323433663731363234316531303666303833353662633637663364346332646333376333613930633036626533336535386335383839376139616661396426726f6c653d66726565" alt="1581382ac041220c1ee7b09b7deeda8a.png" data-canonical-src="https://tc-cdn.flowus.cn/oss/7412b6f1-4de1-4671-88ec-94733c19b316/1581382ac041220c1ee7b09b7deeda8a.png?time=1729486800&amp;token=dd243f716241e106f08356bc67f3d4c2dc37c3a90c06be33e58c58897a9afa9d&amp;role=free" style="max-width: 100%;"></a></p></div>
<div style="font-size:small;margin-top:8px;float:right;"></div>

<button class="btn btn-block" type="button" onclick="openComments()" id="cmButton">评论</button>
<div class="comments" id="comments"></div>

</div>
    <div id="footer"><div id="footer1">Copyright © <span id="copyrightYear"></span> <a href="https://bionoob7.github.io">Ma's bioinfo notebook</a></div>
<div id="footer2">
    <span id="runday"></span><span>Powered by <a href="https://meekdai.com/Gmeek.html" target="_blank">Gmeek</a></span>
</div>

<script>
var now=new Date();
document.getElementById("copyrightYear").innerHTML=now.getFullYear();

if(""!=""){
    var startSite=new Date("");
    var diff=now.getTime()-startSite.getTime();
    var diffDay=Math.floor(diff/(1000*60*60*24));
    document.getElementById("runday").innerHTML="网站运行"+diffDay+"天"+" • ";
}
</script></div>
</body>
<script>
var IconList={'sun': 'M8 10.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5zM8 12a4 4 0 100-8 4 4 0 000 8zM8 0a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0V.75A.75.75 0 018 0zm0 13a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0v-1.5A.75.75 0 018 13zM2.343 2.343a.75.75 0 011.061 0l1.06 1.061a.75.75 0 01-1.06 1.06l-1.06-1.06a.75.75 0 010-1.06zm9.193 9.193a.75.75 0 011.06 0l1.061 1.06a.75.75 0 01-1.06 1.061l-1.061-1.06a.75.75 0 010-1.061zM16 8a.75.75 0 01-.75.75h-1.5a.75.75 0 010-1.5h1.5A.75.75 0 0116 8zM3 8a.75.75 0 01-.75.75H.75a.75.75 0 010-1.5h1.5A.75.75 0 013 8zm10.657-5.657a.75.75 0 010 1.061l-1.061 1.06a.75.75 0 11-1.06-1.06l1.06-1.06a.75.75 0 011.06 0zm-9.193 9.193a.75.75 0 010 1.06l-1.06 1.061a.75.75 0 11-1.061-1.06l1.06-1.061a.75.75 0 011.061 0z', 'moon': 'M9.598 1.591a.75.75 0 01.785-.175 7 7 0 11-8.967 8.967.75.75 0 01.961-.96 5.5 5.5 0 007.046-7.046.75.75 0 01.175-.786zm1.616 1.945a7 7 0 01-7.678 7.678 5.5 5.5 0 107.678-7.678z', 'sync': 'M1.705 8.005a.75.75 0 0 1 .834.656 5.5 5.5 0 0 0 9.592 2.97l-1.204-1.204a.25.25 0 0 1 .177-.427h3.646a.25.25 0 0 1 .25.25v3.646a.25.25 0 0 1-.427.177l-1.38-1.38A7.002 7.002 0 0 1 1.05 8.84a.75.75 0 0 1 .656-.834ZM8 2.5a5.487 5.487 0 0 0-4.131 1.869l1.204 1.204A.25.25 0 0 1 4.896 6H1.25A.25.25 0 0 1 1 5.75V2.104a.25.25 0 0 1 .427-.177l1.38 1.38A7.002 7.002 0 0 1 14.95 7.16a.75.75 0 0 1-1.49.178A5.5 5.5 0 0 0 8 2.5Z', 'home': 'M6.906.664a1.749 1.749 0 0 1 2.187 0l5.25 4.2c.415.332.657.835.657 1.367v7.019A1.75 1.75 0 0 1 13.25 15h-3.5a.75.75 0 0 1-.75-.75V9H7v5.25a.75.75 0 0 1-.75.75h-3.5A1.75 1.75 0 0 1 1 13.25V6.23c0-.531.242-1.034.657-1.366l5.25-4.2Zm1.25 1.171a.25.25 0 0 0-.312 0l-5.25 4.2a.25.25 0 0 0-.094.196v7.019c0 .138.112.25.25.25H5.5V8.25a.75.75 0 0 1 .75-.75h3.5a.75.75 0 0 1 .75.75v5.25h2.75a.25.25 0 0 0 .25-.25V6.23a.25.25 0 0 0-.094-.195Z', 'github': 'M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z', 'copy': 'M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z', 'check': 'M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z'};
var utterancesLoad=0;

let themeSettings={
    "dark": ["dark","moon","#00f0ff","dark-blue"],
    "light": ["light","sun","#ff5000","github-light"],
    "auto": ["auto","sync","","preferred-color-scheme"]
};
function changeTheme(mode, icon, color, utheme){
    document.documentElement.setAttribute("data-color-mode",mode);
    document.getElementById("themeSwitch").setAttribute("d",value=IconList[icon]);
    document.getElementById("themeSwitch").parentNode.style.color=color;
    if(utterancesLoad==1){utterancesTheme(utheme);}
}
function modeSwitch(){
    let currentMode=document.documentElement.getAttribute('data-color-mode');
    let newMode = currentMode === "light" ? "dark" : currentMode === "dark" ? "auto" : "light";
    localStorage.setItem("meek_theme", newMode);
    if(themeSettings[newMode]){
        changeTheme(...themeSettings[newMode]);
    }
}
function utterancesTheme(theme){
    const message={type:'set-theme',theme: theme};
    const iframe=document.getElementsByClassName('utterances-frame')[0];
    iframe.contentWindow.postMessage(message,'https://utteranc.es');
}
if(themeSettings[theme]){changeTheme(...themeSettings[theme]);}
console.log("\n %c Gmeek last https://github.com/Meekdai/Gmeek \n","padding:5px 0;background:#02d81d;color:#fff");
</script>

<script>
document.getElementById("pathHome").setAttribute("d",IconList["home"]);
document.getElementById("pathIssue").setAttribute("d",IconList["github"]);



function openComments(){
    cm=document.getElementById("comments");
    cmButton=document.getElementById("cmButton");
    cmButton.innerHTML="loading";
    span=document.createElement("span");
    span.setAttribute("class","AnimatedEllipsis");
    cmButton.appendChild(span);

    script=document.createElement("script");
    script.setAttribute("src","https://utteranc.es/client.js");
    script.setAttribute("repo","bionoob7/bionoob7.github.io");
    script.setAttribute("issue-term","title");
    
    if(localStorage.getItem("meek_theme")=="dark"){script.setAttribute("theme","dark-blue");}
    else if(localStorage.getItem("meek_theme")=="light") {script.setAttribute("theme","github-light");}
    else{script.setAttribute("theme","preferred-color-scheme");}
    
    script.setAttribute("crossorigin","anonymous");
    script.setAttribute("async","");
    cm.appendChild(script);

    int=self.setInterval("iFrameLoading()",200);
}

function iFrameLoading(){
    var utterances=document.getElementsByClassName('utterances');
    if(utterances.length==1){
        if(utterances[0].style.height!=""){
            utterancesLoad=1;
            int=window.clearInterval(int);
            document.getElementById("cmButton").style.display="none";
            console.log("utterances Load OK");
        }
    }
}

document.addEventListener('DOMContentLoaded', () => {
    const createClipboardHTML = (codeContent, additionalClasses = '') => `
        <pre class="notranslate"><code class="notranslate">${codeContent}</code></pre>
        <div class="clipboard-container position-absolute right-0 top-0 ${additionalClasses}">
            <clipboard-copy class="ClipboardButton btn m-2 p-0" role="button" style="display: inherit;">
                <svg height="16" width="16" class="octicon octicon-copy m-2"><path d="${IconList["copy"]}"></path></svg>
                <svg height="16" width="16" class="octicon octicon-check color-fg-success m-2 d-none"><path d="${IconList["check"]}"></path></svg>
            </clipboard-copy>
            <div class="copy-feedback">Copied!</div>
        </div>
    `;

    const handleCodeElements = (selector = '') => {
        document.querySelectorAll(selector).forEach(codeElement => {
            const codeContent = codeElement.innerHTML;
            const newStructure = document.createElement('div');
            newStructure.className = 'snippet-clipboard-content position-relative overflow-auto';
            newStructure.innerHTML = createClipboardHTML(codeContent);

            const parentElement = codeElement.parentElement;
            if (selector.includes('highlight')) {
                parentElement.insertBefore(newStructure, codeElement.nextSibling);
                parentElement.removeChild(codeElement);
            } else {
                parentElement.parentElement.replaceChild(newStructure, parentElement);
            }
        });
    };

    handleCodeElements('pre.notranslate > code.notranslate');
    handleCodeElements('div.highlight > pre.notranslate');

    let currentFeedback = null;
    document.querySelectorAll('clipboard-copy').forEach(copyButton => {
        copyButton.addEventListener('click', () => {
            const codeContent = copyButton.closest('.snippet-clipboard-content').innerText;
            const tempTextArea = document.createElement('textarea');
            tempTextArea.value = codeContent;
            document.body.appendChild(tempTextArea);
            tempTextArea.select();
            document.execCommand('copy');
            document.body.removeChild(tempTextArea);

            const copyIcon = copyButton.querySelector('.octicon-copy');
            const checkIcon = copyButton.querySelector('.octicon-check');
            const copyFeedback = copyButton.nextElementSibling;

            if (currentFeedback && currentFeedback !== copyFeedback) {currentFeedback.style.display = 'none';}
            currentFeedback = copyFeedback;

            copyIcon.classList.add('d-none');
            checkIcon.classList.remove('d-none');
            copyFeedback.style.display = 'block';
            copyButton.style.borderColor = 'var(--color-success-fg)';

            setTimeout(() => {
                copyIcon.classList.remove('d-none');
                checkIcon.classList.add('d-none');
                copyFeedback.style.display = 'none';
                copyButton.style.borderColor = '';
            }, 2000);
        });
    });
});

</script>
<script src='https://blog.meekdai.com/Gmeek/plugins/GmeekTOC.js'></script><script src='https://blog.meekdai.com/Gmeek/plugins/lightbox.js'></script>

</html>
